[SPARK-13980] Incrementally serialize blocks while unrolling them in MemoryStore

JoshRosen opened the pull request:
When a block is persisted in the MemoryStore at a serialized storage level, the current MemoryStore.putIterator() code will unroll the entire iterator as Java objects in memory, then will turn around and serialize an iterator obtained from the unrolled array. This is inefficient and doubles our peak memory requirements.

Instead, I think that we should incrementally serialize blocks while unrolling them.

A downside to incremental serialization is the fact that we will need to deserialize the partially-unrolled data in case there is not enough space to unroll the block and the block cannot be dropped to disk. However, I'm hoping that the memory efficiency improvements will outweigh any performance losses as a result of extra serialization in that hopefully-rare case.

JoshRosen committed:
Duplicate the code to separate putIterator() by serialization level.

JoshRosen committed:
Incrementally serialize blocks while unrolling them.

SparkQA commented:
**[Test build #53456 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53456/consoleFull)** for PR 11791 at commit [`7dc3623`](https://github.com/apache/spark/commit/7dc362331f3f549670ecd9488db456b4136a3ad7).

SparkQA commented:
**[Test build #53456 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53456/consoleFull)** for PR 11791 at commit [`7dc3623`](https://github.com/apache/spark/commit/7dc362331f3f549670ecd9488db456b4136a3ad7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53456/
Test PASSed.

SparkQA commented:
**[Test build #53496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53496/consoleFull)** for PR 11791 at commit [`5489748`](https://github.com/apache/spark/commit/5489748d7e8b0d4aa7a0e7331a1a4a02f65d977f).

SparkQA commented:
**[Test build #53496 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53496/consoleFull)** for PR 11791 at commit [`5489748`](https://github.com/apache/spark/commit/5489748d7e8b0d4aa7a0e7331a1a4a02f65d977f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53496/
Test PASSed.

JoshRosen commented:
/cc @rxin @andrewor14, this is the next most important patch to review towards off-heap caching. After these changes get in, we'll be able to use off-heap memory for the unroll memory in off-heap caching, greatly simplifying things. Without this change, the on-heap unroll array needs to be accounted properly even if the final cache destination is off-heap, making the caching more OOM-prone and complicating the accounting logic (since it then becomes different between the two modes).

rxin commented:
Still WIP?

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into serialize-incrementally

JoshRosen committed:
Remove local variable to reduce potential for use-after-invalid.

JoshRosen committed:
More docs.

JoshRosen committed:
Roll back unintentional reformatting.

SparkQA commented:
**[Test build #53700 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53700/consoleFull)** for PR 11791 at commit [`a336c17`](https://github.com/apache/spark/commit/a336c177fec2cca5157619524c597032ef2212aa).

JoshRosen commented:
This is no longer WIP and should be ready for review now.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -129,10 +136,9 @@ private[spark] class MemoryStore(
    *         partially-unrolled block.
    */
-  private[storage] def putIterator(
+  private[storage] def putIteratorAsValues(
```
In case it isn't obvious from the diff, the main change in this file is to split `putIterator` into two separate methods, `putIteratorAsValues` and `putIteratorAsBytes`.

It's possible that there's some opportunity to reduce code duplication here, but unless we can come up with an obvious and simple approach I would prefer to defer cleanup to followup patches.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -500,3 +601,79 @@ private[storage] class PartiallyUnrolledIterator(
+    new PartiallyUnrolledIterator(
+      memoryStore,
+      unrollMemory,
+      unrolled = blockManager.dataDeserialize(blockId, unrolled),
```
It's kind of a bummer that we need to pass `blockManager` in here when all that we want to do is figure out which type of decompression and deserialization needs to be performed. It'd be nice to loosen coupling here in a future followup patch.

JoshRosen replied:
Also, note that the changes in this patch are going to moderately conflict with my other patch for auto-selection of caching serializers, since we'll then need to make sure that the proper classTags are propagated here.

SparkQA commented:
**[Test build #53700 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53700/consoleFull)** for PR 11791 at commit [`a336c17`](https://github.com/apache/spark/commit/a336c177fec2cca5157619524c597032ef2212aa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53700/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #53704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53704/consoleFull)** for PR 11791 at commit [`a336c17`](https://github.com/apache/spark/commit/a336c177fec2cca5157619524c597032ef2212aa).

SparkQA commented:
**[Test build #53704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53704/consoleFull)** for PR 11791 at commit [`a336c17`](https://github.com/apache/spark/commit/a336c177fec2cca5157619524c597032ef2212aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53704/
Test PASSed.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -244,13 +244,113 @@ private[spark] class MemoryStore(
+        if (keepUnrolling) {
+          unrollMemoryUsedByThisBlock += amountToRequest
+        }
+        unrollMemoryUsedByThisBlock += amountToRequest
```
i dont understand why you add this twice in some cases

JoshRosen replied:
Ah, I think this case is a mistake which might have been introduced while repairing a merge conflict. We should only increment this if `keepUnrolling == true`.

JoshRosen replied:
This has now been fixed.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -244,13 +244,113 @@ private[spark] class MemoryStore(
+      reserveAdditionalMemoryIfNecessary()
+    }
+
+    if (keepUnrolling) {
```
move this into line 317?

JoshRosen replied:
This is actually here on purpose and deserves a comment. The goal here is to make sure that once we reach line 317 we are guaranteed to have enough memory to store the block. When we finish serializing the block and reach line 311, it's possible that the actual memory usage has exceeded our unroll memory slightly, so here we do one final bumping up of the unroll memory.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -244,13 +244,113 @@ private[spark] class MemoryStore(
+      val entry = SerializedMemoryEntry(
+        new ChunkedByteBuffer(byteArrayChunkOutputStream.toArrays.map(ByteBuffer.wrap)))
+      // Synchronize so that transfer is atomic
+      memoryManager.synchronized {
```
I see this replicated in a few places (transferUnrollToStorage()). It might be easier to have releaseUnrollMemoryForThisTask take another argument to optionally transfer it to storage

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -244,13 +244,113 @@ private[spark] class MemoryStore(
+    }
+
+    def reserveAdditionalMemoryIfNecessary(): Unit = {
+      if (byteArrayChunkOutputStream.size > unrollMemoryUsedByThisBlock) {
```
One important implicit assumption which I will make explicit in a line comment: we assume that we'll always be able to get enough memory to unroll at least one element in between size calculation. This is the same assumption that we have in the deserialized case, since we only periodically measure memory usage there.

nongli commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -1074,7 +1074,7 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
 
     // Unroll with all the space in the world. This should succeed.
-    var putResult = memoryStore.putIterator("unroll", smallList.iterator, StorageLevel.MEMORY_ONLY)
+    var putResult = memoryStore.putIteratorAsValues("unroll", smallList.iterator)
```
Do we need more test cases for putIteratorAsBytes?

JoshRosen replied:
Good call; yes. I'll add some, since there's actually a couple of new branches that need coverage.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -500,3 +601,79 @@ private[storage] class PartiallyUnrolledIterator(
+  /**
+   * Called to dispose of this block and free its memory.
+   */
+  def discard(): Unit = {
```
who calls this now?

JoshRosen replied:
It's not actually called in the current code, since it's only currently called in one place, `BlockManager.doPutIterator()`, which returns a `PartiallyUnrolledIterator`, so that sole callsite ends up calling `valuesIterator()`, which takes care of the discarding.

rxin commented:
cc @sameeragarwal

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into serialize-incrementally

SparkQA commented:
**[Test build #53741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53741/consoleFull)** for PR 11791 at commit [`cec1f02`](https://github.com/apache/spark/commit/cec1f02b36a0a52f977b800847d9a182b2958ff0).

JoshRosen committed:
Remove accidental double-increment.

SparkQA commented:
**[Test build #53742 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53742/consoleFull)** for PR 11791 at commit [`4976b74`](https://github.com/apache/spark/commit/4976b743c8ac8eda86b548ea6a858d9c6f164591).

SparkQA commented:
**[Test build #53742 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53742/consoleFull)** for PR 11791 at commit [`4976b74`](https://github.com/apache/spark/commit/4976b743c8ac8eda86b548ea6a858d9c6f164591).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53742/
Test PASSed.

SparkQA commented:
**[Test build #53741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53741/consoleFull)** for PR 11791 at commit [`cec1f02`](https://github.com/apache/spark/commit/cec1f02b36a0a52f977b800847d9a182b2958ff0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53741/
Test PASSed.

JoshRosen committed:
Explain potentially-confusing if case.

JoshRosen committed:
Pass SerializerManager to MemoryStore constructor.

JoshRosen commented:
Doing a bit of refactoring now in order to make it easier to write proper unit tests for this. Therefore I'd hold off on the final review pass here for a little bit and review my SPARK-3000 patch instead.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into serialize-incrementally

SparkQA commented:
**[Test build #53957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53957/consoleFull)** for PR 11791 at commit [`768a8d9`](https://github.com/apache/spark/commit/768a8d95ef394553ced64c1fad683127e54e699b).

SparkQA commented:
**[Test build #53957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53957/consoleFull)** for PR 11791 at commit [`768a8d9`](https://github.com/apache/spark/commit/768a8d95ef394553ced64c1fad683127e54e699b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LogisticRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class NaiveBayesModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class KMeansModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class Binarizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Bucketizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class CountVectorizer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class CountVectorizerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class DCT(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class ElementwiseProduct(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, JavaMLReadable,`
  * `class IDF(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class IDFModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class MaxAbsScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class MaxAbsScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class MinMaxScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class NGram(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Normalizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class OneHotEncoder(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class PolynomialExpansion(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class QuantileDiscretizer(JavaEstimator, HasInputCol, HasOutputCol, HasSeed, JavaMLReadable,`
  * `class RegexTokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class SQLTransformer(JavaTransformer, JavaMLReadable, JavaMLWritable):`
  * `class StandardScaler(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class StandardScalerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid, JavaMLReadable,`
  * `class StringIndexerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class IndexToString(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class StopWordsRemover(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Tokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorAssembler(JavaTransformer, HasInputCols, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorIndexer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class VectorIndexerModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class VectorSlicer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Word2VecModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class PCA(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class PCAModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class RFormula(JavaEstimator, HasFeaturesCol, HasLabelCol, JavaMLReadable, JavaMLWritable):`
  * `class RFormulaModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class ChiSqSelector(JavaEstimator, HasFeaturesCol, HasOutputCol, HasLabelCol, JavaMLReadable,`
  * `class ChiSqSelectorModel(JavaModel, JavaMLReadable, JavaMLWritable):`
  * `class PipelineMLWriter(JavaMLWriter):`
  * `class Pipeline(Estimator, MLReadable, MLWritable):`
  * `class PipelineModelMLWriter(JavaMLWriter):`
  * `class PipelineModel(Model, MLReadable, MLWritable):`
  * `class ALSModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class LinearRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class IsotonicRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class AFTSurvivalRegressionModel(JavaModel, JavaMLWritable, JavaMLReadable):`
  * `class MLWriter(object):`
  * `class JavaMLWriter(MLWriter):`
  * `class JavaMLWritable(MLWritable):`
  * `class MLReader(object):`
  * `class JavaMLReader(MLReader):`
  * `class JavaMLReadable(MLReadable):`
  * `  implicit class StringToColumn(val sc: StringContext) `
  * `class RecordReaderIterator[T](rowReader: RecordReader[_, T]) extends Iterator[T] `
  * `        // the type in next() and we get a class cast exception.  If we make that function return`
  * `class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String)`
  * `class StreamProgress(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53957/
Test PASSed.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into serialize-incrementally

JoshRosen committed:
Add more tests.

JoshRosen commented:
Alright, just added a few more tests to `MemoryStoreSuite` to bump up the coverage of `putBytesAsIterator()` and fixed a problem leading to leaked unroll memory in `PartiallySerializedResult.finishWritingToStream`.

SparkQA commented:
**[Test build #54057 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54057/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).

SparkQA commented:
**[Test build #54057 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54057/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54057/
Test FAILed.

SparkQA commented:
**[Test build #54070 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54070/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).

SparkQA commented:
**[Test build #54070 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54070/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54070/
Test FAILed.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -279,13 +275,117 @@ private[spark] class MemoryStore(
+
+    // Make sure that we have enough memory to store the block. By this point, it is possible that
+    // the block's actual memory usage has exceeded the unroll memory by a small amount, so we
+    // perform one final call to attempt to allocate additional memory if necessary.
```
This is because of the call to close? That can use more memory?

JoshRosen replied:
Yes.

JoshRosen commented:
Jenkins, retest this please.

nongli commented:
LGTM

SparkQA commented:
**[Test build #54094 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54094/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -876,21 +876,40 @@ private[spark] class BlockManager(
+                size = diskStore.getSize(blockId)
+              } else {
+                iteratorFromFailedMemoryStorePut = Some(partiallySerializedValues.valuesIterator)
+              }
```
duplicate code >:(

andrewor14 replied:
You can clean it up easily here:
```
val putResult: Either[...] =
  if (level.deserialized) {
    memoryStore.putIteratorAsValues(...)
  } else {
    memoryStore.putIteratorAsBytes(...)
  }
putResult match {
  case Right(s) => s
  case Left(partiallySerializedValues) => ...
}
```

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -876,21 +876,40 @@ private[spark] class BlockManager(
+              } else {
+                iteratorFromFailedMemoryStorePut = Some(partiallySerializedValues.valuesIterator)
+              }
+          }
```
about my previous comment about duplicate code, never mind. It can't actually be abstracted cleanly.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -279,13 +275,117 @@ private[spark] class MemoryStore(
+      reserveAdditionalMemoryIfNecessary()
+    }
+
+    if (keepUnrolling) {
```
why have two of these `if`'s? Can't we merge them?

JoshRosen replied:
See the comment above the first one.

andrewor14 replied:
I don't get it. What change in functionality will there be if we moved those 2 lines into this if case? Did you separate them for readability?

JoshRosen replied:
After the `reserveAdditionalMemoryIfNecessary()` call on line 347, `keepUnrolling` can become `false`.

andrewor14 replied:
I see...

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -544,3 +645,81 @@ private[storage] class PartiallyUnrolledIterator[T](
+ */
+private class RedirectableOutputStream extends OutputStream {
+  private[this] var os: OutputStream = _
+  def setOutputStream(s: OutputStream): Unit = { os = s }
```
Do we need a whole new class for this? Can we just write (failed stream + the rest of input stream) to the file when we drop it to disk?

andrewor14 commented:
Looks good.

SparkQA commented:
**[Test build #54094 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54094/consoleFull)** for PR 11791 at commit [`749df73`](https://github.com/apache/spark/commit/749df73c8dee8733a2617ff9ab8563fe3ed9de66).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54094/
Test PASSed.

JoshRosen commented:
Merging to master.
